rig-core/src/agent.rs

```diff
@@ -169,100 +169,118 @@ pub struct Agent<M: CompletionModel> {
+                                    let text = serde_json::to_string_pretty(&doc)
+                                        .unwrap_or_else(|_| doc.to_string());
+
+                                    Document {
+                                        id,
+                                        text,
+                                        additional_props: HashMap::new(),
+                                    }
+                                })
+                                .collect::<Vec<_>>(),
+                        )
+                    })
+                    .try_fold(vec![], |mut acc, docs| async {
+                        acc.extend(docs);
+                        Ok(acc)
+                    })
+                    .await
+                    .map_err(|e| CompletionError::RequestError(Box::new(e)))?;
+
+                let dynamic_tools = stream::iter(self.dynamic_tools.iter())
+                    .then(|(num_sample, index)| async {
+                        Ok::<_, VectorStoreError>(
+                            index
+                                .top_n_ids(text, *num_sample)
+                                .await?
+                                .into_iter()
+                                .map(|(_, id)| id)
+                                .collect::<Vec<_>>(),
+                        )
+                    })
+                    .try_fold(vec![], |mut acc, docs| async {
+                        for doc in docs {
+                            if let Some(tool) = self.tools.get(&doc) {
+                                acc.push(tool.definition(text.into()).await)
+                            } else {
+                                tracing::warn!("Tool implementation not found in toolset: {}", doc);
+                            }
+                        }
+                        Ok(acc)
+                    })
+                    .await
+                    .map_err(|e| CompletionError::RequestError(Box::new(e)))?;
+
+                let static_tools = stream::iter(self.static_tools.iter())
+                    .filter_map(|toolname| async move {
+                        if let Some(tool) = self.tools.get(toolname) {
+                            Some(tool.definition(text.into()).await)
+                        } else {
+                            tracing::warn!(
+                                "Tool implementation not found in toolset: {}",
+                                toolname
+                            );
+                            None
+                        }
+                    })
+                    .collect::<Vec<_>>()
+                    .await;
+
+                completion_request
+                    .documents([self.static_context.clone(), dynamic_context].concat())
+                    .tools([static_tools.clone(), dynamic_tools].concat())
+            }
+            None => completion_request,
```

Comment:
`static_context` and `static_tools` should be added even if `prompt.rag_text() == None`

Reply:
`static_context` works but `static_tools` is difficult bc the prompt is used to look up the tool via the toolset.

```rs
let static_tools = stream::iter(self.static_tools.iter())
    .filter_map(|toolname| async move {
        if let Some(tool) = self.tools.get(toolname) {
            Some(tool.definition(text.into()).await)
        } else {
            tracing::warn!("Tool implementation not found in toolset: {}", toolname);
            None
        }
    })
    .collect::<Vec<_>>()
    .await;
```